---
layout: single
classes: wide
title: "우당탕탕 1"
categories:
  - 우당탕탕
last_modified_at: 2020-02-07T13:00:00+09:00
---


써본 언어라고는 통계 분석을 위한 Python 과 R 뿐인 무지랭이이므로 천천히 진행을 해본다.


``` go
package main

import (
	"fmt"
	"io/ioutil"
)

func main() {
	file, err := ioutil.ReadFile("/Users/seonwoolim/Desktop/대학원/2학기/데마/data/harris.dat")
	fmt.Println(string(file))
}
```

구글링으로 찾아와서 그대로 따라한 이 코드의 뜻을 대충 짐작해본다.
* import야 뭐 패키지 가져온 것이고, io/ioutil 이 편해보여서 가져와봤다.
* main이라는 것이 내가 만들 무언가..인 것 같다.
* ioutil의 ReadFile 함수로 harris.dat 파일을 가져왔다. 주의할 점은, **ReadFile 함수의 output은 두개라는 점이다**.
두번째 output은 error 이다. 사람에 따라 error 이 없지 않을 경우 (`if err != nil `) 에러 메시지를 출력하거나 프로그램이 멈춰버리도록(`panic`,`log.Fatal` 등) 코드를 추가한 것을 봤지만 나는 귀찮아서 안했다.
* print를 해보기 위해서 string 으로 변환하였다. 이 부분은 추후에 없어질 예정이다.

코드를 test.go 라는 이름으로 저장한다. R이나 Python은 한 줄씩 실행이 되지만 go 는 그렇지가 않다.
심지어 코드 짜다 말고 저장하면 **이 변수는 왜 declare만 하고 쓰진 않은거야!** 하고 나한테 화를 낸다.
그러니 중간에 코드를 잘 짰는지 보고 싶다면 대충 코드를 마무리한 척 하고 돌려보도록 하자.

아무튼 저런 코드를 쓰고 저장하면 그냥 test.go 파일이 하나 만들어져 있을 뿐이다.
이 놈을 실행하려면 터미널에서 저 파일이 있는 디렉토리로 간 다음에 go build test.go 하면 된다.
그러면 드디어 실행!

이 아니고 같은 디렉토리에 exe 파일이 생긴다. ~~개빡쳐~~ 그걸 실행시켜야 비로소 나의 코드가 실행되는 것이다. 아 참으로 신묘하다.
결과물은 아래와 같이 예쁘게 프린트가 되더라.


```
Last login: Fri Jan 24 12:37:03 on ttys001
(base) seonui-iMac:~ seonwoolim$ /Users/seonwoolim/Desktop/test ; exit;
3900 12 0 1 0
4020 10 44 7 0
4290 12 5 30 0
4380 8 6.2 7 0
4380 8 7.5 6 0
4380 12 0 7 0
#길어서 중간은 생략
6600 15 64 16 1
6600 15 84 33 1
6600 15 215.5 16 1
6840 15 41.5 7 1
6900 12 175 10 1
6900 15 132 24 1
8100 16 54.5 33 1
logout
Saving session...
...copying shared history...
...saving history...truncating history files...
...completed.

[프로세스 완료됨]

```

...Gophernote 를 쓰기로 결정했다. 매번 이 짓을 할 순 없어... 설치는 링크 https://github.com/gopherdata/gophernotes 를 참고했다. ~~그리고 수많은 오류들과 싸우게 되는데...~~ (덧 : 결국 이 무지랭이는 오늘 gophernote 설치를 포기하게 됩니다.)

여기서 참고사항 : 나는 mac osx 를 사용하고 있다. 그리고 컴알못이라 그저 시키는대로 흘러가는대로 해낼 뿐이다. 무슨 일이 있었는지 하나하나 쓰는 이유는? 1. 혹시나 gophernote를 쓰려는 mac 이용자에게 요만큼이라도 도움이 될까 싶어서 2. 내가 도움 받으려고 (두번째 이유가 더 큼). 아무튼 링크의 첫번째 줄을 터미널에 실행시켰다.

```
$ env GO111MODULE=on go get github.com/gopherdata/gophernotes
```

그랬더니 pkg-config 가 없어서 안된다고 한다. ~~그게뭔데~~ 그래서 설치를 했다. 설치 방법은 이 링크 참고 : http://macappstore.org/pkg-config

그리고 다시 위의 명령어를 실행시키니까, **또 오류가 난다!** 이번엔 libzmq 라는 패키지가 없다고 한다. ~~그게 뭔데!!~~ 설치 방법은 그냥 `brew install zmq` 하면 된다.

기다렸더니 잘 돌아갔다. **휴!** 이제 다 된 거겠지? 나머지 줄들을 실행시켜보자.

```
$mkdir -p ~/Library/Jupyter/kernels/gophernotes
$cd ~/Library/Jupyter/kernels/gophernotes
$cp "$(go env GOPATH)"/pkg/mod/github.com/gopherdata/gophernotes@v0.6.1/kernel/*  "."

```

결과는 오류.

```
cp: ./kernel.json: Permission denied
cp: ./kernel.json.in: Permission denied
cp: ./logo-32x32.png: Permission denied
cp: ./logo-64x64.png: Permission denied
```
...라고 뜬다. 왜냐하면 이새끼가 날 얕보고 있기 때문이다. 기계덩어리에게 인간의 권력을 보여주자. 마지막으로 썼던 명령어 맨 앞에 sudo 를 붙이기만 하면 된다. 이제 마지막 줄만 남았다.

```
$sed "s|gophernotes|$(go env GOPATH)/bin/gophernotes|" < kernel.json.in > kernel.json
```
실행했더니 역시나 permission denied. sudo로 혼내줘도 permission denied. 아아악! 이렇게 하랬잖아요! 찾아보니 누군가가 저기 kernel.json.in 을 감싸고 있는 뾰족괄호들을 없애서 해결했다. (sudo 는 앞에 붙인 상태)그래서 나도 해봤더니, 됐다! 뾰족괄호가 뭔 역할을 하는진 몰라도 아무튼 됐다! 설치가 잘 됐는지 확인하기 위해 다음의 명령을 실행시켜보자.

```
$ "$(go env GOPATH)"/bin/gophernotes
```
오늘의 날짜와 시간, 그리고 Need a command line argument specifying the connection file. 이라는 문장이 나오면 잘 설치된 것이라고 한다. 아이 행복해! 이제 Gophernote를 사용해볼까?

...Kernel Error이 난다. 아 세상에.... -to be continued-

```
	Traceback (most recent call last):
  File "/Users/seonwoolim/anaconda3/lib/python3.7/site-packages/tornado/web.py", line 1699, in _execute
    result = await result
  File "/Users/seonwoolim/anaconda3/lib/python3.7/site-packages/tornado/gen.py", line 742, in run
    yielded = self.gen.throw(*exc_info)  # type: ignore
  File "/Users/seonwoolim/anaconda3/lib/python3.7/site-packages/notebook/services/sessions/handlers.py", line 72, in post
    type=mtype))
  File "/Users/seonwoolim/anaconda3/lib/python3.7/site-packages/tornado/gen.py", line 735, in run
    value = future.result()
  File "/Users/seonwoolim/anaconda3/lib/python3.7/site-packages/tornado/gen.py", line 742, in run
    yielded = self.gen.throw(*exc_info)  # type: ignore
  File "/Users/seonwoolim/anaconda3/lib/python3.7/site-packages/notebook/services/sessions/sessionmanager.py", line 88, in create_session
    kernel_id = yield self.start_kernel_for_session(session_id, path, name, type, kernel_name)
  File "/Users/seonwoolim/anaconda3/lib/python3.7/site-packages/tornado/gen.py", line 735, in run
    value = future.result()
  File "/Users/seonwoolim/anaconda3/lib/python3.7/site-packages/tornado/gen.py", line 742, in run
    yielded = self.gen.throw(*exc_info)  # type: ignore
  File "/Users/seonwoolim/anaconda3/lib/python3.7/site-packages/notebook/services/sessions/sessionmanager.py", line 101, in start_kernel_for_session
    self.kernel_manager.start_kernel(path=kernel_path, kernel_name=kernel_name)
  File "/Users/seonwoolim/anaconda3/lib/python3.7/site-packages/tornado/gen.py", line 735, in run
    value = future.result()
  File "/Users/seonwoolim/anaconda3/lib/python3.7/site-packages/tornado/gen.py", line 209, in wrapper
    yielded = next(result)
  File "/Users/seonwoolim/anaconda3/lib/python3.7/site-packages/notebook/services/kernels/kernelmanager.py", line 168, in start_kernel
    super(MappingKernelManager, self).start_kernel(**kwargs)
  File "/Users/seonwoolim/anaconda3/lib/python3.7/site-packages/jupyter_client/multikernelmanager.py", line 110, in start_kernel
    km.start_kernel(**kwargs)
  File "/Users/seonwoolim/anaconda3/lib/python3.7/site-packages/jupyter_client/manager.py", line 259, in start_kernel
    **kw)
  File "/Users/seonwoolim/anaconda3/lib/python3.7/site-packages/jupyter_client/manager.py", line 204, in _launch_kernel
    return launch_kernel(kernel_cmd, **kw)
  File "/Users/seonwoolim/anaconda3/lib/python3.7/site-packages/jupyter_client/launcher.py", line 138, in launch_kernel
    proc = Popen(cmd, **kwargs)
  File "/Users/seonwoolim/anaconda3/lib/python3.7/subprocess.py", line 775, in __init__
    restore_signals, start_new_session)
  File "/Users/seonwoolim/anaconda3/lib/python3.7/subprocess.py", line 1522, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'gophernotes': 'gophernotes'

```
